### PR TITLE
Support for multiple block devices using fsusermount

### PR DESCRIPTION
**Pull request by dpgeorge**
This brings multiple block device support for extmod/fsusermount.  Stmhal is updated to work with the new code.  The flash and SD card are now exposed as block devices using the block protocol, and mounted as such.

You can now mount up to 4 devices (easy to increase this number at compile time).  You can unmount the flash and SD card and remount them.

Once this lands we can probably move stmhal/diskio.c and stmhal/ffconf.c to lib/utils (or elsewhere outside stmhal) because they should now be fully generic.

Also, a lot of the functions in stmhal/moduos.c are generic to anything using the fatfs library.

### CODE REVIEW

File: stmhal/sdcard.c [hunk 1]
```diff
@@ -346,22 +358,82 @@ STATIC mp_obj_t sd_write(mp_obj_t self, mp_obj_t block_num, mp_obj_t data) {
...
+            return MP_OBJ_NEW_SMALL_INT(0); // success
+
+        case 3: // SYNC
+            // nothing to do
+            return MP_OBJ_NEW_SMALL_INT(0); // success
+
+        case 4: // SEC_COUNT
+            return MP_OBJ_NEW_SMALL_INT(0); // TODO
+
+        case 5: // SEC_SIZE
```
Comment (pfalcon) [inline stmhal/sdcard.c]: So, there're now BP_IOCTL_* constants for these, so they should be used.

**Comment by pfalcon** [comment]
> The maximum allowed is fixed at compile time by the size of the fs_user_mount array accessed via MP_STATE_PORT.

Not exactly. This number is bound by MICROPY_FATFS_VOLUMES as propagated as _VOLUMES to FatFs. I was getting really funny memory trashing when made similar mistake. So, should just use MICROPY_FATFS_VOLUMES where you have 4 now, and bump it.

**Comment by pfalcon** [comment]
Otherwise, looks good, thanks!

File: stmhal/sdcard.c (hunk 1 above)
Comment (dpgeorge) [inline stmhal/sdcard.c]: True.  But they need to be moved out of stmhal/diskio.c and into a header.  I guess extmod/fsusermount.h?

**Comment by dpgeorge** [comment]
> Not exactly. This number is bound by MICROPY_FATFS_VOLUMES as propagated as _VOLUMES to FatFs. I was getting really funny memory trashing when made similar mistake.

Yes, thanks for that!  I was also getting strange errors but they seemed to go away when I updated the VolToPart structure to include a 4th entry.  But I now see that definitely the FatFs array can trash memory if not the right size.

Ok, I'll use MICROPY_FATFS_VOLUMES to set the limit on the number of block devices.  But one will have to be careful to update VolToPart as well.  (We could actually get rid of VolToPart and replace it with a simple macro, because the data it holds can be computed from the fs_user_mount_t structure.)

File: stmhal/sdcard.c (hunk 1 above)
Comment (pfalcon) [inline stmhal/sdcard.c]: > True.  But they need to be moved out of stmhal/diskio.c and into a
> header.  I guess extmod/fsusermount.h?

Yes, there, if you don't feel like making a bdev.h in some dir (extmod
is good enough).

Comment (dpgeorge) [inline stmhal/sdcard.c]: I don't feel like making another header file right now.  fsusermount.h has the ioctl method so makes sense to define the ioctl commands in that file.

**Comment by dpgeorge** [comment]
Merged in b33a7705966a743c5e4c0c3c3bb83e6d97b5e84d, and BP_IOCTL constants moved to fsusermount.h in e372e83b3042cb48fbe5706522d92f7aa5404ae6.

**Pull request closed without merging**